File: frontend/src/types/preparation.ts
export type SagaStepType =
  | "VEHICLE_RESERVATION"
  | "TRAJECTORY_PLANNING"
  | "RESOURCE_VALIDATION"
  | "COMMUNICATION_PROFILE"
  | "SIMULATION_INITIALIZATION";

export type SagaStepStatus =
  | "PENDING"
  | "IN_PROGRESS"
  | "COMPLETED"
  | "FAILED"
  | "COMPENSATING"
  | "COMPENSATED";

export interface SagaStep {
  id: string;
  saga_id: string;
  mission_id: string;

  step_type: SagaStepType;
  status: SagaStepStatus;

  failure_reason: string | null;

  started_at: string | null;
  completed_at: string | null;
  failed_at: string | null;
  compensated_at: string | null;
}

export interface MissionPreparation {
  mission_id: string;
  saga_id: string | null;
  steps: SagaStep[];
}

File: services/mission-service/app/schemas/preparation.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.domain.enums import SagaStepStatus, SagaStepType


class SagaStepResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    saga_id: UUID
    mission_id: UUID

    step_type: SagaStepType
    status: SagaStepStatus

    failure_reason: str | None

    started_at: datetime | None
    completed_at: datetime | None
    failed_at: datetime | None
    compensated_at: datetime | None


class MissionPreparationResponse(BaseModel):
    mission_id: UUID
    saga_id: UUID | None
    steps: list[SagaStepResponse]
